nova/modules/core/helpers/Nova_extension_helper.php

```diff
@@ -0,0 +1,209 @@
+					$path = APPPATH.'extensions/'.$this->name.'/views/'.$section.'/pages/'.$view.'.php';
+					if(file_exists($path)){
+            $output = $ci->load->_ci_load(array(
+              '_ci_view' => $location, 
+              '_ci_vars' => $ci->load->_ci_object_to_array($data), 
+              '_ci_path' => $path,
+              '_ci_return' => true
+            ));
+						$ci->event->trigger(['location', 'view', 'output', $obj->sec, $obj->view], [
+							'data' => &$data,
+							'output' => &$output
+						]);
+            return $output;
+          }
+				}
+				
+				$output = $ci->nova->view($location, $data, true);
+				$ci->event->trigger(['location', 'view', 'output', $obj->sec, $obj->view], [
+					'data' => &$data,
+					'output' => &$output
+				]);
+				return $output;
+			}
+			else
+			{
+				$output = $ci->load->view($location, $data, true);
+				$ci->event->trigger(['location', 'view', $obj->sec, $obj->view, 'output'], [
+					'data' => &$data,
+					'output' => &$output
+				]);
+				return $output;
+			}
+		}
+		
+		return $location;
+	}
+
+  public function offsetSet($offset, $value)
+	{
+    if(is_null($offset))
+		{
+      show_error('Extension append by integer not allowed');
+    }
+		else
+		{
+      $this->facades[$offset] = $value;
+    }
+  }
+
+  public function offsetExists($offset)
+	{
+    return isset($this->facades[$offset]);
+  }
+
+  public function offsetUnset($offset)
+	{
+    unset($this->facades[$offset]);
+  }
+
+  public function offsetGet($offset)
+	{
+    return isset($this->facades[$offset]) ? $this->facades[$offset] : null;
+  }
+}
+
+/**
+ * Object $this scope for the extension's init.php file, enabling calls such as
+ * $this->attach('example_lib', new MyExtensionExampleLib()) and 
+ * $this->require('other_extension_name') from within init.php. Besides these
+ * special functions, all other calls and parameter access are forwarded to 
+ * the CodeIgniter controller returned by {@see get_instance()}.
+ */
+class Nova_extension_initialization_context {
+	
+	public $ci;
+	
+	protected $_extensionContainer;
+	
+	public function __construct($extension, $name)
+	{
+		$this->name = $name;
+		$this->path = APPPATH.'extensions/'.$name.'/';
+		$this->ci =& get_instance();
+		$this->_extensionContainer =& $extension;
+	}
+	
+	// Call-forward to the CI context
+	public function __get($name)
+	{
+		return $this->ci->$name;
+	}
+	
+	// Call-forward to the CI context
+	public function __call($name, $args)
+	{
+		return call_user_func_array([$this->ci, $name], $args);
+	}
+	 
+	public function initialize()
+	{
+		// init.php run with $this context of Nova_extension_initialization_context
+		if(file_exists($this->path.'init.php'))
+		{
+			require_once($this->path.'init.php');
+		}
+	}
+	
+	public function attach($name, $object)
+	{
+		$this->_extensionContainer[$name] = $object;
+	}
+	
+	public function require_extension($extension)
+	{
+		$this->_extensionContainer->required_extensions[] = $extension;
+	}
+	
+}
```

Comment:
This doesn't really belong in `helpers`. Helpers are simple procedural functions that get called throughout. Since this is a full-blown class, we need to find another home for it.

Reply:
I didn't like `libraries` for these because they're used internally by the event library and it wouldn't make sense to allow them to be included by `->load->library()` since that brings a singleton instance of the class into the context `$ci->LIBRARY_NAME`. I stuck them in helpers because they really are helpers that support the library.